symengine/sets.cpp

```diff
@@ -393,7 +416,52 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
     if (is_a<UniversalSet>(*o) or is_a<EmptySet>(*o) or is_a<Union>(*o)) {
         return (*o).set_intersection(rcp_from_this_cast<const Set>());
     }
-    throw std::runtime_error("Not implemented");
+    return SymEngine::set_intersection({rcp_from_this_cast<const Set>(), o},
+                                       false);
+}
+
+RCP<const Set> FiniteSet::set_complement(const RCP<const Set> &o) const
+{
+    if (is_a<FiniteSet>(*o)) {
+        const FiniteSet &other = down_cast<const FiniteSet &>(*o);
+        set_basic container;
+        std::set_difference(other.container_.begin(), other.container_.end(),
+                            container_.begin(), container_.end(),
+                            std::inserter(container, container.begin()),
+                            RCPBasicKeyLess{});
+        return finiteset(container);
+    }
+
+    if (is_a<Interval>(*o)) {
+        if (o->compare(*interval(NegInf, Inf, true, true)) == 0) {
```

Comment:
this would be faster than the current implementation, where you are taking complement with `[-oo, oo]` and then intersection with the actual `universe`, with just a couple of checks.